File: threads/src/main/java/im/threads/activities/FilesActivity.java
package im.threads.activities;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.graphics.PorterDuff;
import android.os.Build;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.support.annotation.Nullable;
import android.support.v4.content.ContextCompat;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.KeyEvent;
import android.view.View;
import android.view.Window;
import android.view.WindowManager;
import android.view.inputmethod.EditorInfo;
import android.view.inputmethod.InputMethodManager;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.TextView;

import java.util.Iterator;
import java.util.List;

import im.threads.R;
import im.threads.adapters.FilesAndMediaAdapter;
import im.threads.controllers.FilesAndMediaController;
import im.threads.model.ChatStyle;
import im.threads.model.FileDescription;
import im.threads.utils.PrefUtils;

/**
 * Created by yuri on 01.07.2016.
 */
public class FilesActivity extends BaseActivity implements FilesAndMediaAdapter.OnFileClick {
    private static final String TAG = "FilesActivity ";
    private FilesAndMediaController mFilesAndMediaController;
    private RecyclerView mRecyclerView;
    private EditText mSearchEditText;
    private Toolbar mToolbar;
    private FilesAndMediaAdapter mFilesAndMediaAdapter;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        final Context ctx = this;
        super.onCreate(savedInstanceState);
        ChatStyle style = PrefUtils.getIncomingStyle(this);
        if (Build.VERSION.SDK_INT > 20) {
            Window window = getWindow();
            window.clearFlags(WindowManager.LayoutParams.FLAG_TRANSLUCENT_STATUS);
            window.addFlags(WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS);
            if (style != null) {
                window.setStatusBarColor(getResources().getColor(style.chatStatusBarColorResId));
            }
        }
        setContentView(R.layout.activity_files_and_media);
        mToolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(mToolbar);
        mToolbar.setNavigationOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                onBackPressed();
            }
        });
        if (getFragmentManager().findFragmentByTag(TAG) == null) {
            mFilesAndMediaController = FilesAndMediaController.getInstance();
            getFragmentManager().beginTransaction().add(mFilesAndMediaController, TAG).commit();
        } else {
            mFilesAndMediaController = (FilesAndMediaController) getFragmentManager().findFragmentByTag(TAG);
        }
        mFilesAndMediaController.bindActivity(this);
        mFilesAndMediaController.getFilesAcync();
        mRecyclerView = (RecyclerView) findViewById(R.id.recycler);
        mRecyclerView.setLayoutManager(new LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false));
        mSearchEditText = (EditText) findViewById(R.id.search_edit_text);
        findViewById(R.id.search).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (mSearchEditText.getVisibility() == View.VISIBLE) {
                    mSearchEditText.setText("");
                    mSearchEditText.setVisibility(View.GONE);
                    mToolbar.setTitle(getString(R.string.threads_files_and_media));
                    if (null != mFilesAndMediaAdapter) mFilesAndMediaAdapter.undoClear();
                } else {
                    mSearchEditText.setVisibility(View.VISIBLE);
                    mSearchEditText.requestFocus();
                    mToolbar.setTitle("");
                    if (null != mFilesAndMediaAdapter) mFilesAndMediaAdapter.backupAndClear();
                    mSearchEditText.setText("");
                    new Handler(Looper.getMainLooper()).postDelayed(new Runnable() {
                        @Override
                        public void run() {
                            InputMethodManager imm = (InputMethodManager) getSystemService(INPUT_METHOD_SERVICE);
                            imm.showSoftInput(mSearchEditText, InputMethodManager.SHOW_IMPLICIT);
                        }
                    }, 100);
                }
            }
        });
        mSearchEditText.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
            }

            @Override
            public void afterTextChanged(Editable s) {
                String str = null;
                if (s == null) {
                    str = "";
                } else {
                    str = s.toString();
                }
                search(str);
            }
        });

        mSearchEditText.setOnEditorActionListener(new TextView.OnEditorActionListener() {
            @Override
            public boolean onEditorAction(TextView v, int actionId, KeyEvent event) {

                if ((mSearchEditText.getVisibility() == View.VISIBLE) && actionId == EditorInfo.IME_ACTION_SEARCH) {
                    search(v.getText().toString());
                    return true;
                } else {
                    return false;
                }
            }
        });


        setActivityStyle(PrefUtils.getIncomingStyle(this));
    }

    protected void search(String searchString) {
        if (mFilesAndMediaAdapter != null) {
            mFilesAndMediaAdapter.filter(searchString);
        }
    }

    @Override
    protected void setActivityStyle(ChatStyle style) {
        if (null != style) {
            findViewById(R.id.activity_root).setBackgroundColor(ContextCompat.getColor(this, style.filesAndMediaScreenBackgroundColor));

            mToolbar.setBackgroundColor(ContextCompat.getColor(this, style.chatToolbarColorResId));

            ((ImageButton) findViewById(R.id.search)).setColorFilter(ContextCompat.getColor(this, style.chatToolbarTextColorResId), PorterDuff.Mode.SRC_ATOP);
            mSearchEditText.setTextColor(getColorInt(style.chatToolbarTextColorResId));
            mToolbar.getNavigationIcon().setColorFilter(ContextCompat.getColor(this, style.chatToolbarTextColorResId), PorterDuff.Mode.SRC_ATOP);

            mSearchEditText.setHintTextColor(getColorInt(style.chatToolbarHintTextColor));
        }
    }

    public void onFileReceive(List<FileDescription> descriptions) {
        if (descriptions != null && descriptions.size() > 0) {
            for (Iterator<FileDescription> iter = descriptions.iterator(); iter.hasNext(); ) {
                if (iter.next().getFilePath() == null) iter.remove();
            }
            mFilesAndMediaAdapter = new FilesAndMediaAdapter(descriptions, this);
            mRecyclerView.setAdapter(mFilesAndMediaAdapter);

        }
    }

    @Override
    public void onFileClick(FileDescription fileDescription) {
        mFilesAndMediaController.onFileClick(fileDescription);
    }

    public static Intent getStartIntetent(Activity activity) {
        Intent i = new Intent(activity, FilesActivity.class);
        return i;
    }

    @Override
    public void onBackPressed() {
        if (mSearchEditText.getVisibility() == View.VISIBLE) {
            mSearchEditText.setText("");
            mSearchEditText.setVisibility(View.GONE);
            mToolbar.setTitle(getString(R.string.threads_files_and_media));
            if (null != mFilesAndMediaAdapter) mFilesAndMediaAdapter.undoClear();
        } else {
            super.onBackPressed();
        }
    }
}

File: threads/src/main/java/im/threads/ui/activities/filesActivity/FilesActivity.kt
package im.threads.ui.activities.filesActivity

import android.app.Activity
import android.content.ActivityNotFoundException
import android.content.Intent
import android.graphics.PorterDuff
import android.graphics.Typeface
import android.os.Build
import android.os.Bundle
import android.text.Editable
import android.text.TextUtils
import android.text.TextWatcher
import android.util.TypedValue
import android.view.Gravity
import android.view.KeyEvent
import android.view.View
import android.view.WindowManager
import android.view.inputmethod.EditorInfo
import android.widget.TextView
import androidx.activity.viewModels
import androidx.annotation.ColorRes
import androidx.annotation.DimenRes
import androidx.annotation.StringRes
import androidx.core.content.ContextCompat
import im.threads.R
import im.threads.business.models.FileDescription
import im.threads.business.secureDatabase.DatabaseHolder
import im.threads.business.serviceLocator.core.inject
import im.threads.business.utils.Balloon
import im.threads.databinding.EccActivityFilesAndMediaBinding
import im.threads.ui.ChatStyle
import im.threads.ui.activities.BaseActivity
import im.threads.ui.adapters.filesAndMedia.FilesAndMediaAdapter
import im.threads.ui.adapters.filesAndMedia.FilesAndMediaAdapter.OnFileClick
import im.threads.ui.config.Config
import im.threads.ui.utils.ColorsHelper
import im.threads.ui.utils.gone
import im.threads.ui.utils.hideKeyboard
import im.threads.ui.utils.setColorFilter
import im.threads.ui.utils.showKeyboard
import im.threads.ui.utils.visible

/**
 * Показывает список файлов, которые присутствовали в диалоге с оператором с обоих сторон
 */
internal class FilesActivity : BaseActivity(), OnFileClick {
    private val binding: EccActivityFilesAndMediaBinding by lazy {
        EccActivityFilesAndMediaBinding.inflate(layoutInflater)
    }
    private val config: Config by lazy {
        Config.getInstance()
    }
    private val database: DatabaseHolder by inject()
    private val filesViewModel: FilesViewModel by viewModels {
        FilesViewModel.Factory(config.context, database)
    }
    private var filesAndMediaAdapter: FilesAndMediaAdapter? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        hideSystemUI()
        setStatusBarColor()
        setContentView(binding.root)
        initToolbar()
        setOnSearchClickAction()
        setOnSearchTextChanged()
        setActivityStyle(config.getChatStyle())
        subscribeForNewIntents()
        subscribeForDownloadProgress()
        requestFiles()
    }

    override fun onBackPressed() {
        if (binding.searchEditText.visibility == View.VISIBLE) {
            closeSearch()
        } else {
            super.onBackPressed()
        }
    }

    override fun onFileClick(fileDescription: FileDescription?) {
        filesViewModel.onFileClick(fileDescription)
    }

    override fun onDownloadFileClick(fileDescription: FileDescription?) {
        filesViewModel.onDownloadFileClick(fileDescription)
    }

    private fun onFileReceive(descriptions: List<FileDescription?>?) = with(binding) {
        if (descriptions != null && descriptions.isNotEmpty()) {
            searchButton.visibility = View.VISIBLE
            emptyListLayout.visibility = View.GONE
            filesRecycler.visibility = View.VISIBLE
            filesAndMediaAdapter = FilesAndMediaAdapter(
                descriptions,
                this@FilesActivity,
                this@FilesActivity
            )
            filesRecycler.adapter = filesAndMediaAdapter
        }
    }

    @Suppress("DEPRECATION")
    private fun hideSystemUI() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            window.setDecorFitsSystemWindows(true)
        } else {
            window.clearFlags(WindowManager.LayoutParams.FLAG_TRANSLUCENT_STATUS)
        }
        window.addFlags(WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS)
    }

    private fun setStatusBarColor() {
        val statusBarColor = ContextCompat.getColor(baseContext, config.getChatStyle().mediaAndFilesStatusBarColorResId)
        val isStatusBarLight = resources.getBoolean(config.getChatStyle().mediaAndFilesWindowLightStatusBarResId)
        super.setStatusBarColor(isStatusBarLight, statusBarColor)
    }

    private fun initToolbar() = with(binding) {
        setSupportActionBar(toolbar)
        toolbar.setNavigationOnClickListener { onBackPressed() }
        val isShadowVisible = resources.getBoolean(config.getChatStyle().isChatTitleShadowVisible)
        toolbarShadow.visibility = if (isShadowVisible) View.VISIBLE else View.INVISIBLE
        if (!isShadowVisible) {
            val noElevation = 0f
            toolbar.elevation = noElevation
        }
        ColorsHelper.setTint(this@FilesActivity, backButton, config.getChatStyle().chatToolbarTextColorResId)
        title.setTextColor(ContextCompat.getColor(this@FilesActivity, config.getChatStyle().chatToolbarTextColorResId))
        initToolbarTextPosition()
        setClickForBackBtn()
    }

    private fun initToolbarTextPosition() = with(binding) {
        val isToolbarTextCentered = Config.getInstance().getChatStyle().isToolbarTextCentered
        val gravity = if (isToolbarTextCentered) Gravity.CENTER else Gravity.CENTER_VERTICAL
        title.gravity = gravity
    }

    private fun setClickForBackBtn() {
        binding.backButton.setOnClickListener {
            onBackPressed()
        }
    }

    private fun setOnSearchClickAction() = with(binding) {
        searchButton.setOnClickListener {
            if (searchEditText.visibility == View.VISIBLE) {
                closeSearch()
            } else {
                showSearch()
            }
        }
    }

    private fun closeSearch() = with(binding) {
        searchEditText.setText("")
        searchEditText.hideKeyboard()
        searchEditText.gone()
        title.visible()
        filesAndMediaAdapter?.undoClear()
    }

    private fun showSearch() = with(binding) {
        searchEditText.visible()
        searchEditText.requestFocus()
        title.gone()
        filesAndMediaAdapter?.backupAndClear()
        searchEditText.setText("")
        searchEditText.showKeyboard(100)
    }

    private fun setOnSearchTextChanged() = with(binding) {
        searchEditText.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(text: CharSequence, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(text: CharSequence, start: Int, before: Int, count: Int) {}
            override fun afterTextChanged(text: Editable) {
                search(text.toString())
            }
        })
        searchEditText.setOnEditorActionListener { textView: TextView, actionId: Int, _: KeyEvent? ->
            if (searchEditText.visibility == View.VISIBLE && actionId == EditorInfo.IME_ACTION_SEARCH) {
                search(textView.text.toString())
                return@setOnEditorActionListener true
            } else {
                return@setOnEditorActionListener false
            }
        }
    }

    private fun requestFiles() {
        filesViewModel.getFilesAsync()
    }

    private fun search(searchString: String) = with(binding) {
        filesAndMediaAdapter?.filter(searchString) {
            if (filesAndMediaAdapter?.isNotEmpty() == true) {
                filesRecycler.scrollToPosition(0)
            }
        }
    }

    private fun setActivityStyle(style: ChatStyle) = with(binding) {
        root.setBackgroundColor(
            ContextCompat.getColor(
                baseContext,
                style.mediaAndFilesScreenBackgroundColor
            )
        )
        toolbar.setBackgroundColor(
            ContextCompat.getColor(
                baseContext,
                style.mediaAndFilesToolbarColorResId
            )
        )
        searchButton.setColorFilter(
            ContextCompat.getColor(
                baseContext,
                style.mediaAndFilesToolbarTextColorResId
            ),
            PorterDuff.Mode.SRC_ATOP
        )
        searchEditText.setTextColor(getColorInt(style.mediaAndFilesToolbarTextColorResId))
        toolbar.navigationIcon?.mutate()?.setColorFilter(
            ContextCompat.getColor(baseContext, style.mediaAndFilesToolbarTextColorResId)
        )
        searchEditText.setHintTextColor(getColorInt(style.mediaAndFilesToolbarHintTextColor))
        setUpTextViewStyle(
            emptyListHeader,
            style.emptyMediaAndFilesHeaderText,
            style.emptyMediaAndFilesHeaderTextSize,
            style.emptyMediaAndFilesHeaderTextColor,
            style.emptyMediaAndFilesHeaderFontPath
        )
        setUpTextViewStyle(
            emptyListDescription,
            style.emptyMediaAndFilesDescriptionText,
            style.emptyMediaAndFilesDescriptionTextSize,
            style.emptyMediaAndFilesDescriptionTextColor,
            style.emptyMediaAndFilesDescriptionFontPath
        )
    }

    private fun subscribeForNewIntents() {
        filesViewModel.intentLiveData.observe(this) { startIntent(it) }
    }

    private fun subscribeForDownloadProgress() {
        filesViewModel.filesFlowLiveData.observe(this) { onDownloadProgress(it) }
    }

    private fun setUpTextViewStyle(
        textView: TextView,
        @StringRes textResId: Int,
        @DimenRes sizeResId: Int,
        @ColorRes colorResId: Int,
        fontPath: String?
    ) = with(binding) {
        textView.setText(textResId)
        textView.setTextSize(TypedValue.COMPLEX_UNIT_PX, resources.getDimension(sizeResId))
        textView.setTextColor(getColorInt(colorResId))
        if (!TextUtils.isEmpty(fontPath)) {
            textView.typeface = Typeface.createFromAsset(assets, fontPath)
        }
    }

    private fun startIntent(intent: Intent) {
        try {
            startActivity(intent)
        } catch (e: ActivityNotFoundException) {
            Balloon.show(this, getString(R.string.ecc_file_not_supported))
        }
    }

    private fun onDownloadProgress(filesFlow: FilesFlow) {
        when (filesFlow) {
            is FilesFlow.UpdatedProgress -> {
                filesAndMediaAdapter?.updateProgress(filesFlow.fileDescription)
            }
            is FilesFlow.DownloadError -> {
                filesAndMediaAdapter?.onDownloadError(filesFlow.fileDescription)
            }
            is FilesFlow.FilesReceived -> {
                onFileReceive(filesFlow.files)
            }
        }
    }

    companion object {
        /**
         * Запускает FilesActivity
         * @param activity активити, из которой будет произведен запуск FilesActivity.
         */
        @JvmStatic
        fun startActivity(activity: Activity?) {
            activity?.startActivity(Intent(activity, FilesActivity::class.java))
        }
    }
}
